File: python-sdk/observability/metrics/pusher.py
"""Kafka metrics pusher."""

import os
import sys
import threading
import time
from typing import Optional

import grpc
from prometheus_client import REGISTRY
from prometheus_client.core import (
    CounterMetricFamily,
    GaugeMetricFamily,
    HistogramMetricFamily,
    SummaryMetricFamily,
)

from opentelemetry.proto.collector.metrics.v1.metrics_service_pb2 import (
    ExportMetricsServiceRequest,
)
from opentelemetry.proto.common.v1.common_pb2 import AnyValue, KeyValue
from opentelemetry.proto.metrics.v1.metrics_pb2 import (
    AggregationTemporality,
    Gauge,
    Histogram,
    HistogramDataPoint,
    Metric,
    NumberDataPoint,
    ResourceMetrics,
    ScopeMetrics,
    Sum,
    Summary,
    SummaryDataPoint,
)
from opentelemetry.proto.resource.v1.resource_pb2 import Resource

from opentelemetry.proto.collector.metrics.v1.metrics_service_pb2_grpc import (
    MetricsServiceStub,
)

from observability.metrics.registry import Registry


class MetricsPusher:
    """Periodically pushes Prometheus metrics to an OTEL collector via OTLP gRPC."""
    
    def __init__(
        self,
        service_name: str,
        otlp_endpoint: str,
        interval: float,
    ):
        self.service_name = service_name
        self.otlp_endpoint = otlp_endpoint
        self.interval = interval
        self._channel: Optional[grpc.Channel] = None
        self._stub: Optional[MetricsServiceStub] = None
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None
        
        self._init_client()
    
    def _init_client(self):
        """Initialize OTLP gRPC client."""
        try:
            self._channel = grpc.insecure_channel(self.otlp_endpoint)
            self._stub = MetricsServiceStub(self._channel)
        except Exception as e:
            sys.stderr.write(f"Failed to initialize OTLP metrics client: {e}\n")
            self._channel = None
            self._stub = None
    
    def start(self):
        """Start the background push thread."""
        if self._stub:
            self._thread = threading.Thread(target=self._push_loop, daemon=True)
            self._thread.start()
    
    def stop(self):
        """Stop the pusher and flush remaining metrics."""
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=5.0)
        if self._channel:
            self._channel.close()
    
    def _push_loop(self):
        """Background loop to push metrics."""
        while not self._stop_event.wait(self.interval):
            self._push()
    
    def _push(self):
        """Push current metrics to OTEL collector in OTLP protobuf format."""
        if not self._stub:
            return
        
        try:
            # Convert Prometheus metrics to OTLP format
            otlp_request = self._convert_to_otlp()
            
            # Send via OTLP gRPC
            self._stub.Export(otlp_request)
        except Exception as e:
            sys.stderr.write(f"Failed to push metrics: {e}\n")
    
    def _convert_to_otlp(self) -> ExportMetricsServiceRequest:
        """Convert Prometheus metrics to OTLP format."""
        timestamp_nanos = int(time.time() * 1e9)
        metrics = []
        
        # Iterate through metric families
        for collector in Registry._collector_to_names.keys():
            for metric_family in collector.collect():
                metric_name = metric_family.name
                metric_type = metric_family.type
                metric_help = metric_family.documentation
                
                # Create metric based on type
                if metric_type == "counter":
                    for sample in metric_family.samples:
                        attributes = [
                            KeyValue(
                                key=label_name,
                                value=AnyValue(string_value=str(label_value))
                            )
                            for label_name, label_value in sample.labels.items()
                        ]
                        metric = Metric(
                            name=sample.name,
                            description=metric_help,
                            sum=Sum(
                                aggregation_temporality=AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE,
                                is_monotonic=True,
                                data_points=[
                                    NumberDataPoint(
                                        attributes=attributes,
                                        time_unix_nano=timestamp_nanos,
                                        as_double=float(sample.value),
                                    )
                                ],
                            ),
                        )
                        metrics.append(metric)
                
                elif metric_type == "gauge":
                    for sample in metric_family.samples:
                        attributes = [
                            KeyValue(
                                key=label_name,
                                value=AnyValue(string_value=str(label_value))
                            )
                            for label_name, label_value in sample.labels.items()
                        ]
                        metric = Metric(
                            name=sample.name,
                            description=metric_help,
                            gauge=Gauge(
                                data_points=[
                                    NumberDataPoint(
                                        attributes=attributes,
                                        time_unix_nano=timestamp_nanos,
                                        as_double=float(sample.value),
                                    )
                                ],
                            ),
                        )
                        metrics.append(metric)
                
                elif metric_type == "histogram":
                    # Group samples by label set (excluding 'le')
                    groups: dict[str, dict] = {}
                    for sample in metric_family.samples:
                        labels_copy = {k: v for k, v in sample.labels.items() if k != "le"}
                        key = str(sorted(labels_copy.items()))
                        
                        if key not in groups:
                            groups[key] = {
                                "labels": labels_copy,
                                "bucket_bounds": [],
                                "bucket_counts": [],
                                "count": 0,
                                "sum_value": 0.0,
                            }
                        group = groups[key]
                        
                        if sample.name.endswith("_bucket"):
                            le = sample.labels.get("le", "")
                            if le != "+Inf":
                                group["bucket_bounds"].append(float(le))
                                group["bucket_counts"].append(int(sample.value))
                            else:
                                # +Inf bucket — overflow count
                                group["bucket_counts"].append(int(sample.value))
                        elif sample.name.endswith("_count"):
                            group["count"] = int(sample.value)
                        elif sample.name.endswith("_sum"):
                            group["sum_value"] = float(sample.value)
                    
                    for group in groups.values():
                        attributes = [
                            KeyValue(
                                key=label_name,
                                value=AnyValue(string_value=str(label_value))
                            )
                            for label_name, label_value in group["labels"].items()
                        ]
                        metric = Metric(
                            name=metric_name,
                            description=metric_help,
                            histogram=Histogram(
                                aggregation_temporality=AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE,
                                data_points=[
                                    HistogramDataPoint(
                                        attributes=attributes,
                                        time_unix_nano=timestamp_nanos,
                                        count=group["count"],
                                        sum=group["sum_value"],
                                        bucket_counts=group["bucket_counts"],
                                        explicit_bounds=group["bucket_bounds"],
                                    )
                                ],
                            ),
                        )
                        metrics.append(metric)
                
                elif metric_type == "summary":
                    # Group samples by label set (excluding 'quantile')
                    groups = {}
                    for sample in metric_family.samples:
                        labels_copy = {k: v for k, v in sample.labels.items() if k != "quantile"}
                        key = str(sorted(labels_copy.items()))
                        
                        if key not in groups:
                            groups[key] = {
                                "labels": labels_copy,
                                "quantiles": [],
                                "count": 0,
                                "sum_value": 0.0,
                            }
                        group = groups[key]
                        
                        if sample.name.endswith("_count"):
                            group["count"] = int(sample.value)
                        elif sample.name.endswith("_sum"):
                            group["sum_value"] = float(sample.value)
                        elif "quantile" in sample.labels:
                            group["quantiles"].append(
                                SummaryDataPoint.ValueAtQuantile(
                                    quantile=float(sample.labels["quantile"]),
                                    value=float(sample.value),
                                )
                            )
                    
                    for group in groups.values():
                        attributes = [
                            KeyValue(
                                key=label_name,
                                value=AnyValue(string_value=str(label_value))
                            )
                            for label_name, label_value in group["labels"].items()
                        ]
                        metric = Metric(
                            name=metric_name,
                            description=metric_help,
                            summary=Summary(
                                data_points=[
                                    SummaryDataPoint(
                                        attributes=attributes,
                                        time_unix_nano=timestamp_nanos,
                                        count=group["count"],
                                        sum=group["sum_value"],
                                        quantile_values=group["quantiles"],
                                    )
                                ],
                            ),
                        )
                        metrics.append(metric)
        
        # Create the OTLP request
        return ExportMetricsServiceRequest(
            resource_metrics=[
                ResourceMetrics(
                    resource=Resource(
                        attributes=[
                            KeyValue(
                                key="service.name",
                                value=AnyValue(string_value=self.service_name)
                            )
                        ]
                    ),
                    scope_metrics=[
                        ScopeMetrics(metrics=metrics)
                    ],
                )
            ]
        )



_pusher: Optional[MetricsPusher] = None


def start_metrics_pusher(service_name: str) -> bool:
    """Start the metrics pusher.
    
    Returns True if started, False if disabled or failed.
    """
    global _pusher
    
    # Skip in development mode
    if _get_env("ENVIRONMENT", "production") == "development":
        return False
    
    interval_str = os.getenv("METRICS_PUSH_INTERVAL", "15")
    
    # Parse interval (remove 's' suffix if present)
    try:
        interval = float(interval_str.rstrip("s"))
    except ValueError:
        interval = 15.0
    
    # Determine OTLP endpoint (collector agent)
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317")
    if "://" in endpoint:
        endpoint = endpoint.split("://", 1)[1]

    _pusher = MetricsPusher(service_name, endpoint, interval)
    _pusher.start()
    
    return True


def stop_metrics_pusher():
    """Stop the metrics pusher."""
    global _pusher
    if _pusher:
        _pusher.stop()
        _pusher = None


def _get_env(key: str, default: str) -> str:
    """Get environment variable with fallback names for ENVIRONMENT."""
    if key == "ENVIRONMENT":
        for k in ["ENVIRONMENT", "ENV", "environment", "env"]:
            v = os.getenv(k)
            if v:
                return v
    
    return os.getenv(key, default)

File: go-sdk/pkg/observability/metrics/pusher.go
package metrics

import (
	"context"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	dto "github.com/prometheus/client_model/go"
	"go.opentelemetry.io/proto/otlp/collector/metrics/v1"
	commonpb "go.opentelemetry.io/proto/otlp/common/v1"
	metricspb "go.opentelemetry.io/proto/otlp/metrics/v1"
	resourcepb "go.opentelemetry.io/proto/otlp/resource/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/proto"
)

// Pusher pushes metrics to an OTEL collector via OTLP gRPC periodically.
type Pusher struct {
	conn        *grpc.ClientConn
	client      v1.MetricsServiceClient
	registry    *prometheus.Registry
	serviceName string
	endpoint    string
	interval    time.Duration
	cancel      context.CancelFunc
	wg          sync.WaitGroup
}

// PusherConfig holds pusher configuration.
type PusherConfig struct {
	ServiceName string
	Endpoint    string
	Interval    time.Duration
}

// NewPusherConfig creates config from environment (mirrors Python metrics pusher).
func NewPusherConfig(serviceName string) *PusherConfig {
	intervalStr := getEnv("METRICS_PUSH_INTERVAL", "15s")
	interval, err := time.ParseDuration(intervalStr)
	if err != nil {
		interval = 15 * time.Second
	}

	endpoint := getEnv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317")
	// Strip scheme if present (e.g. http://host:4317 -> host:4317)
	if strings.Contains(endpoint, "://") {
		parts := strings.SplitN(endpoint, "://", 2)
		if len(parts) == 2 {
			endpoint = parts[1]
		}
	}

	return &PusherConfig{
		ServiceName: serviceName,
		Endpoint:    endpoint,
		Interval:    interval,
	}
}

// NewPusher creates a new metrics pusher backed by OTLP gRPC.
func NewPusher(cfg *PusherConfig) (*Pusher, error) {
	if cfg.Endpoint == "" {
		return nil, nil
	}

	conn, err := grpc.NewClient(cfg.Endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}

	client := v1.NewMetricsServiceClient(conn)

	return &Pusher{
		conn:        conn,
		client:      client,
		registry:    Registry,
		serviceName: cfg.ServiceName,
		endpoint:    cfg.Endpoint,
		interval:    cfg.Interval,
	}, nil
}

// Start begins the background push loop.
func (p *Pusher) Start() {
	ctx, cancel := context.WithCancel(context.Background())
	p.cancel = cancel

	p.wg.Add(1)
	go p.pushLoop(ctx)
}

// Stop gracefully stops the pusher and closes the gRPC connection.
func (p *Pusher) Stop() error {
	if p.cancel != nil {
		p.cancel()
	}
	p.wg.Wait()
	if p.conn != nil {
		return p.conn.Close()
	}
	return nil
}

func (p *Pusher) pushLoop(ctx context.Context) {
	defer p.wg.Done()

	ticker := time.NewTicker(p.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			// Final push before exit
			p.push()
			return
		case <-ticker.C:
			p.push()
		}
	}
}

func (p *Pusher) push() {
	if p.client == nil {
		return
	}

	// Gather metrics from Prometheus registry
	mfs, err := p.registry.Gather()
	if err != nil {
		os.Stderr.WriteString("Failed to gather metrics: " + err.Error() + "\n")
		return
	}

	// Convert to OTLP format
	otlpMetrics := p.convertToOTLP(mfs)
	if otlpMetrics == nil {
		return
	}

	// Marshal for debugging/logging if needed (not required for gRPC call)
	_, err = proto.Marshal(otlpMetrics)
	if err != nil {
		os.Stderr.WriteString("Failed to marshal OTLP metrics: " + err.Error() + "\n")
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if _, err := p.client.Export(ctx, otlpMetrics); err != nil {
		os.Stderr.WriteString("Failed to push metrics via OTLP: " + err.Error() + "\n")
	}
}

// convertToOTLP converts Prometheus metrics to OTLP format
func (p *Pusher) convertToOTLP(mfs []*dto.MetricFamily) *v1.ExportMetricsServiceRequest {
	now := time.Now()
	timeNanos := uint64(now.UnixNano())

	var metrics []*metricspb.Metric

	for _, mf := range mfs {
		for _, m := range mf.GetMetric() {
			metric := &metricspb.Metric{
				Name:        mf.GetName(),
				Description: mf.GetHelp(),
			}

			// Convert labels to attributes
			var attributes []*commonpb.KeyValue
			for _, label := range m.GetLabel() {
				attributes = append(attributes, &commonpb.KeyValue{
					Key: label.GetName(),
					Value: &commonpb.AnyValue{
						Value: &commonpb.AnyValue_StringValue{
							StringValue: label.GetValue(),
						},
					},
				})
			}

			// Convert based on metric type
			switch mf.GetType() {
			case dto.MetricType_COUNTER:
				metric.Data = &metricspb.Metric_Sum{
					Sum: &metricspb.Sum{
						AggregationTemporality: metricspb.AggregationTemporality_AGGREGATION_TEMPORALITY_CUMULATIVE,
						IsMonotonic:            true,
						DataPoints: []*metricspb.NumberDataPoint{
							{
								Attributes:        attributes,
								TimeUnixNano:      timeNanos,
								Value:             &metricspb.NumberDataPoint_AsDouble{AsDouble: m.GetCounter().GetValue()},
							},
						},
					},
				}

			case dto.MetricType_GAUGE:
				metric.Data = &metricspb.Metric_Gauge{
					Gauge: &metricspb.Gauge{
						DataPoints: []*metricspb.NumberDataPoint{
							{
								Attributes:   attributes,
								TimeUnixNano: timeNanos,
								Value:        &metricspb.NumberDataPoint_AsDouble{AsDouble: m.GetGauge().GetValue()},
							},
						},
					},
				}

			case dto.MetricType_HISTOGRAM:
				hist := m.GetHistogram()
				var bucketCounts []uint64
				var explicitBounds []float64

				for _, bucket := range hist.GetBucket() {
					bucketCounts = append(bucketCounts, bucket.GetCumulativeCount())
					explicitBounds = append(explicitBounds, bucket.GetUpperBound())
				}

				metric.Data = &metricspb.Metric_Histogram{
					Histogram: &metricspb.Histogram{
						AggregationTemporality: metricspb.AggregationTemporality_AGGREGATION_TEMPORALITY_CUMULATIVE,
						DataPoints: []*metricspb.HistogramDataPoint{
							{
								Attributes:     attributes,
								TimeUnixNano:   timeNanos,
								Count:          hist.GetSampleCount(),
								Sum:            func() *float64 { v := hist.GetSampleSum(); return &v }(),
								BucketCounts:   bucketCounts,
								ExplicitBounds: explicitBounds,
							},
						},
					},
				}

			case dto.MetricType_SUMMARY:
				summary := m.GetSummary()
				var quantiles []*metricspb.SummaryDataPoint_ValueAtQuantile

				for _, q := range summary.GetQuantile() {
					quantiles = append(quantiles, &metricspb.SummaryDataPoint_ValueAtQuantile{
						Quantile: q.GetQuantile(),
						Value:    q.GetValue(),
					})
				}

				metric.Data = &metricspb.Metric_Summary{
					Summary: &metricspb.Summary{
						DataPoints: []*metricspb.SummaryDataPoint{
							{
								Attributes:     attributes,
								TimeUnixNano:   timeNanos,
								Count:          summary.GetSampleCount(),
								Sum:            summary.GetSampleSum(),
								QuantileValues: quantiles,
							},
						},
					},
				}
			}

			metrics = append(metrics, metric)
		}
	}

	return &v1.ExportMetricsServiceRequest{
		ResourceMetrics: []*metricspb.ResourceMetrics{
			{
				Resource: &resourcepb.Resource{
					Attributes: []*commonpb.KeyValue{
						{
							Key: "service.name",
							Value: &commonpb.AnyValue{
								Value: &commonpb.AnyValue_StringValue{
									StringValue: p.serviceName,
								},
							},
						},
					},
				},
				ScopeMetrics: []*metricspb.ScopeMetrics{
					{
						Metrics: metrics,
					},
				},
			},
		},
	}
}

func getEnv(key, defaultValue string) string {
	if key == "ENVIRONMENT" {
		for _, k := range []string{"ENVIRONMENT", "ENV", "environment", "env"} {
			if v := os.Getenv(k); v != "" {
				return v
			}
		}
	}
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getEnvBool(key string, defaultValue bool) bool {
	value := os.Getenv(key)
	if value == "" {
		return defaultValue
	}
	return value == "true" || value == "1" || value == "yes"
}
